python/runtime/cudaq/algorithms/py_state.cpp

```diff
@@ -134,6 +134,50 @@ void bindPyState(py::module &mod) {
         return pyGetState(kernel, args);
       },
       "Return the state generated by the given quantum kernel.");
+
+  py::class_<async_state_result>(
+      mod, "AsyncStateResult",
+      R"#(A data-type containing the results of a call to :func:`get_state_async`. 
+The `AsyncStateResult` models a future-like type, whose 
+:class:`State` may be returned via an invocation of the `get` method. This 
+kicks off a wait on the current thread until the results are available.
+See `future <https://en.cppreference.com/w/cpp/thread/future>`_ 
+for more information on this programming pattern.)#")
+      .def(
+          "get", [](async_state_result &self) { return self.get(); },
+          "Return the :class:`State` from the asynchronous `get_state` "
+          "accessor execution.\n");
+
+  mod.def(
+      "get_state_async",
+      [](kernel_builder<> &kernel, py::args args, std::size_t qpu_id) {
+        // Ensure the user input is correct.
+        auto validatedArgs = validateInputArguments(kernel, args);
+        auto &platform = cudaq::get_platform();
+        kernel.jitCode();
+        return cudaq::details::runGetStateAsync(
+            [&, a = std::move(validatedArgs)]() mutable {
+              OpaqueArguments argData;
+              packArgs(argData, a);
```

Comment:
May need to be careful here with regards to the GIL and truly releasing it, and having a py::args type in this callable. In my current Python work I create a raw OpaqueArguments pointer and pack it outside of the callable and pass the pointer by value, and delete it at the end of the callable. You might also need to release the GIL at some point (although I'm not sure it will affect this, the Python thread should return immediately.

Reply:
Actually, I was thinking about the GIL when writing this :-) I've updated the code to pack the `OpaqueArguments` outside the callable as suggested.